### PR TITLE
enhancement(config): Allow JSON and YAML config formats in addition to TOML

### LINKED ISSUE #1327: Allow other configuration file formats besides toml
please allow configuration in other formats besides toml. We avoid using software based on toml-files. Especially for complex configuration, the format is imho terrible.

### LINKED ISSUE #3174: Support configuration in JSON
Vector configuration file(s) can get very large and so difficult to look through and maintain. I would like to generate the vector configuration, e.g. by using some configuration languages like [dhall](https://dhall-lang.org/) or [jsonnet](https://jsonnet.org/). Most of the configuration languages can generate JSON but not (yet) TOML. It would be helpful if the `--config` flag supports files in JSON format, as well.

### PR DESCRIPTION
**Pull request by MOZGIII**
This PR adds support for YAML and JSON config formats, in addition to TOML.

Closes #3174.
Closes #1327.

To do:

- [x] behaviour tests
- [x] docs (1)
  - https://github.com/timberio/vector/blob/bc2533d07c6c2b5828040819f625289491205ac4/docs/reference/cli.cue#L113-L117
  - https://vector.dev/docs/setup/configuration/
- [x] docs (2) - add new CLI flag docs and env vars

Not this time:

- generate examples in JSON/YAML

### CODE REVIEW

File: src/config/format.rs
Comment (Hoverbear) [inline src/config/format.rs:12]: Since the fallback is TOML, do you think this is necessary?

Comment (MOZGIII) [reply src/config/format.rs:12]: It allows us to iteratively try and detect the format in multiple steps, and we only fallback at the very end.

There was a discussion about detecting the format based on the contents - I think it will be very useful for that case. Or for other extensions to this format detection process.

**Comment by binarylogic** [comment]
Very nice. I am not the person to review this code, but happy to see it done.

File: src/config/format.rs
Comment (Hoverbear) [inline src/config/format.rs:83]: The default is TOML no? Why do we need these tests for this format unknown?

File: src/config/format.rs
Comment (Hoverbear) [reply src/config/format.rs:12]: I do not see anything in #3174 about supporting files which are not postfixed with `.json`, `.toml`, or `.yaml`.

In that issue @lukesteensen even requested a flag to configure when json was parsed. In that issue yaml was not discussed at all. While I do not think a flag is neccessary, or that supporting yaml is a bad idea, those are discussions that should be had before a merge. :)

File: src/config/format.rs
Comment (MOZGIII) [reply src/config/format.rs:83]: These tests are for the function that determines the format from the file extension.

There can be other functions like this, and the flow would look something like this:
1. When we have the file name (the relevant context), attempt to detect the format for the first time and remember it.
2. When we have the file content (different context), if the file format is still unknown - attempt to detect the file format again from the content, and preserve the format.
3. When we get to actually using the format - fallback to TOML if still unknown, and apply deserialization for the requested format.

File: src/config/format.rs
Comment (MOZGIII) [reply src/config/format.rs:12]: Yep, I'm fine with this! Let's move this discussion to the regular PR comments stream then.

**Comment by MOZGIII** [comment]
Follow up from https://github.com/timberio/vector/pull/5005#discussion_r526417954:

1. YAML: for or against?
2. For/against flag to specify the format?
3. Do we need to attempt to detect the format from contents (vim `mode` annotations etc)

File: src/config/format.rs
Comment (Hoverbear) [reply src/config/format.rs:83]: Since we only support toml, yaml, or json, perhaps attempting to open a file with a different extension should result in an error for now. If we add support for new file types that can remove the error for that type.

**Comment by Hoverbear** [comment]
I think supporting yaml is easy and makes sense here.

I think we can derive file type from the file name. So a flag is not needed.

I do not think we should attempt to parse a file with a file type we don't support. I'd rather not try to be clever or magical, instead instructing the user to name their files appropriately.

**Comment by MOZGIII** [comment]
I agree.

Using a single flag to switch the format is not so great, as users won't be able to mix the files in different formats - which is possible with file-ext-based detection.

**Review by lukesteensen — COMMENTED** [review]
My original thought was simply to add a `--config-json` flag or similar as an alternative to `--config`, just to avoid ambiguity and make the change explicitly opt-in. 

The main issue I see with this approach is that it requires the user to name their files in a particular way. For example, I don't think there's a way to get vector to parse the config as JSON if you do something like the following:

```
$ vector --config <(jsonnet my_config.jsonnet)
``` 

Not that I expect an enormous number of users to do that, but I would prefer we not limit ourselves arbitrarily.

**Comment by lukesteensen** [comment]
> What if we add `--config-toml`, `--config-yaml` and `--config-json` flags that would skip the extension-based detection and assume the per-flag-hardcoded format?

This sounds good to me! I assume we'd leave `--config` to assume TOML? I'm not entirely convinced we need `--config-toml` also, but probably makes sense for consistency.

**Comment by MOZGIII** [comment]
> > What if we add `--config-toml`, `--config-yaml` and `--config-json` flags that would skip the extension-based detection and assume the per-flag-hardcoded format?
> 
> This sounds good to me! I assume we'd leave `--config` to assume TOML? I'm not entirely convinced we need `--config-toml` also, but probably makes sense for consistency.

We can keep the `--config` and make it detect the config format from the file extension. This is great for use cases where we actually want to detect the format from the file extension - like when we pass the `configs/*` glob and the configs are populated, for instance, from the numerous `ConfigMap`s.

**Comment by lukesteensen** [comment]
> We can keep the `--config` and make it detect the config format from the file extension (in addition to the format-specific commads).

I'm slightly hesitant to do this since it's technically not backwards compatible. I'm also not sure how often users will want to use multiple files with mixed formats.

**Comment by MOZGIII** [comment]
> I'm slightly hesitant to do this since it's technically not backwards compatible. I'm also not sure how often users will want to use multiple files with mixed formats.

It is mostly backward compatible - only people having TOML files with `.yaml`/`.json` extensions will hit the new behavior. That is if we keep the fallback.

I wanted to rely on this in the K8s as a "poor man's CRD alternative" to support some use cases easily.

**Comment by MOZGIII** [comment]
I've added the cli flags for each of our commands to pass the configs in a user-specified format. As of now, we support both detecting formats from the file name (very useful when wildcards are used, and files can be in various formats - I expect people will use this a lot in k8s), as well as specifying the exact format to use via one of the format-specific cli flags instead of the generic `--config` one.

File: src/config/format.rs
Comment (Hoverbear) [reply src/config/format.rs:12]: I think we discussed removing this variant?

File: src/config/loading.rs
Comment (Hoverbear) [inline src/config/loading.rs:13]: This seems like it should be toml

File: src/config/loading.rs
Comment (Hoverbear) [inline src/config/loading.rs:18]: Also toml?

File: src/config/format.rs
Comment (Hoverbear) [inline src/config/format.rs:43]: Why not `impl From<AsRef<Path>> for Format`?

File: src/config/format.rs
Comment (Hoverbear) [inline src/config/format.rs:31]: Why not store `Option<Format>` and use `unwrap_or`?

File: src/config/format.rs
Comment (MOZGIII) [reply src/config/format.rs:12]: We talked about generating an error when the format is unknown - and this I didn't do yet to preserve backward compatibility. We can do this in 0.12.
Removing this variant is a bit different - it would be eliminating the whole notion of delayed format resolution (unless we switch to `Option<Format>` which is also fine for the job) - which I think was also discussed, and I think we settled on keeping the overall idea. Or didn't we? Sorry if I've missed anything.

File: src/config/format.rs
Comment (MOZGIII) [reply src/config/format.rs:31]: I thought about this too after the last refactoring. I think we can switch to `Option<Format>`. The downside is it'd be a bit more verbose and thus harder to read, the upside is it's going to be more familiar to people.
It'll also expose the composition of the `Format` (i.e. allow it to be passed without `Option<T>` wrapper) which I don't want here - my thinking was that we want to encapsulate all the logic of the final fallback at the `deserialize` call (i.e. scope it to the `format` mod, and make sure it doesn't escape). But I can tell this is a minor concern.

Given all this, do you think it makes sense to switch to `Option`? I kind of like how it is now, but it'll be fine either way.

File: src/config/loading.rs
Comment (MOZGIII) [reply src/config/loading.rs]: Right, fixed!

File: src/config/format.rs
Comment (MOZGIII) [reply src/config/format.rs]: Indeed, fixed!

**Comment by binarylogic** [comment]
@MOZGIII @Hoverbear we need to get this in 0.11. Are any of these issues blocking?

**Comment by Hoverbear** [comment]
This is pending @lukesteensen 's approval.

**Comment by MOZGIII** [comment]
I think the approach I've implemented here has all of the properties we discussed that we wanted:

- flags for specifying the config format explicitly
- a flag that allows autodetection of the config format from the file path
- mixing and matching of multiple config formats
- backward compat of the `--config` option to a degree

It looks like the only questionable property is that the `--config` is not entirely backwards compatible, because it performs format detection based on the file extension - we repurposed it and added some functionality. We still treat unknown extensions as TOML.

From the code layout perspective and the implementation point of view, a questionable thing is a design that allows lazy format resolution. It just allows for a more complicated logic, that we don't curretly implement but that was discussed at some point.

**Comment by jamtur01** [comment]
I know @Hoverbear has taken a look - I'd really like a second set of eyes on this given the significance of the change. @ktff or @bruceg do you have a moment to take a deeper look?

**Comment by ktff** [comment]
To achieve backward compatibility we can retain current behavior of `--config` flag which assumes TOML, and add `--config-auto` which would detect format based on extension. It's opt in so we can use magic if deemed necessary, and we can then later see how much of usage `--config-auto` gets and then decide if we want to extend it to `--config`.

**Comment by jamtur01** [comment]
I like @ktff's suggestion. +1.

**Comment by binarylogic** [comment]
I'm failing to understand the concern of breaking backward compatibility here. Am I missing something? The only way this would change behavior is if a user supplies a `*.json` o `*.yaml` file that includes TOML, right? Why would anyone do that?

File: src/config/format.rs
Comment (ktff) [reply src/config/format.rs:12]: I also would prefer `Option<Format>`. It's easier to reason about if in all cases we resolved/assigned default format.

**Review by ktff — APPROVED** [review]
That's the only case, a technicality. I don't see any valid reason why would someone name TOML a .json and vice versa. There may be someone who uses it like that, or other corner cases, but I'm not sure that we need to support it.

I'm fine with `--config` doing resolution by extension.

Other than that, PR is good, there is the `Format::Unknown` vs `Option<Format>` dilemma but it's not blocking.

**Comment by binarylogic** [comment]
Yes, and to clarify. `--config` without an extension would default to TOML. So in Luke's example:

```
$ vector --config <(jsonnet my_config.jsonnet)
```

That would evaluate as TOML and error. They would need to use the `--config-json` flag.

**Comment by MOZGIII** [comment]
Ok, I think the path is clear! I'll switch to `Option<Format>` then and merge. No further changes to `--config` - that'll be left as is.
Thanks everyone for the feedback!

**Comment by binarylogic** [comment]
> Ok, I think the path is clear! I'll switch to Option<Format> then and merge. No further changes to --config - that'll be left as is.

@MOZGIII does this mean the following will or will not work?

‘’’
vector —config vector.yaml
‘’’

**Comment by MOZGIII** [comment]
`vector --config vector.yaml` does work in this PR currently and will remain working in `master` and further on.

**Comment by binarylogic** [comment]
That’s a yaml extension, not toml.

**Comment by MOZGIII** [comment]
> That’s a yaml extension, not toml, which does not work on master.

I meant after the merge.

`vector --config qq.yaml` will interpret the config file as YAML (detecting it), just like `vector --config-yaml qq.yaml` (no detection, YAML is assumed).

**Comment by binarylogic** [comment]
Ok great. I was confused by your comment and thought that autodetecting was not enabled. I’m happy that it is.

To be extra explicit. If a user creates a YAML config file and passes it like such, it should work:

vector —config vector.yaml

**Comment by MOZGIII** [comment]
> Ok great. I was confused by your comment and thought that autodetecting was not enabled. I’m happy that it is.
> 
> To be extra explicit. If a user creates a YAML config file and passes it like such, it should work:
> 
> vector —config vector.yaml

Yes, this is correct.

**Comment by MOZGIII** [comment]
Btw, I'm pretty sure that accepting config from std is not currently supported, so this example isn't a backward compat concern:

```shell
$ vector --config <(jsonnet my_config.jsonnet)
```

The reason why I think this is not supported currently is that we have an `stdin` source, and it's not clear how those two would work together in the first place. Also, I don't think I saw the code that would implement this in our codebase while extending the loading routines with formats support.

**Comment by binarylogic** [comment]
@MOZGIII is `vector help` also updated? I wasn't sure where to find that in the code.

**Comment by jamtur01** [comment]
@binarylogic The help text uses the comments in `cli.rs`.

**Comment by MOZGIII** [comment]
> @MOZGIII is `vector help` also updated? I wasn't sure where to find that in the code.

Yep

**Review by ktff — APPROVED** [review]

**Comment by lukesteensen** [comment]
> Btw, I'm pretty sure that accepting config from stdin is not currently supported, so this example isn't a backward compat concern:

That's not what's happening there. That's [process substitution](https://tldp.org/LDP/abs/html/process-sub.html), so vector would see something like `/dev/fd/42` as the argument.

**Comment by MOZGIII** [comment]
Got it! In that case it's also fine - we're reading this path as TOML just as we would before the change. People would want to use `vector --config-json <(jsonnet my_config.jsonnet)` to make it work.

**Pull request merged**
